File: C++/32.服务定位器模式/ServiceLocatorPatternDemo.cpp
/**
 * @file ServiceLocatorPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-12-10
 *
 * @copyright Copyright (c) 2019
 *
 */

#include "ServiceLocator.h"

int main(int argc, char const *argv[]) {
    ServiceLocator service_locator;
    std::shared_ptr<Service> service = service_locator.GetService("Service1");
    service->Execute();
    service = service_locator.GetService("Service2");
    service->Execute();
    service = service_locator.GetService("Service1");
    service->Execute();
    service = service_locator.GetService("Service2");
    service->Execute();

    return 0;
}

File: Objective-C/32.ServiceLocatorPattern/ServiceLocatorPatternDemo.m
//
//  ServiceLocatorPatternDemo.m
//  32.ServiceLocatorPattern
//
//  Created by Layne on 2020/11/22.
//

#import "ServiceLocator.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        id<ServiceProtocol> service = [ServiceLocator getService:@"Service1"];
        [service execute];

        service = [ServiceLocator getService:@"Service2"];
        [service execute];

        service = [ServiceLocator getService:@"Service1"];
        [service execute];

        service = [ServiceLocator getService:@"Service2"];
        [service execute];
    }
    return 0;
}
